controller/gateway/controller_cleanup.go

```diff
@@ -54,68 +54,75 @@ func (r *Reconciler) cleanup(
 	// for cleaning up owned controlplanes when they disappeared.
 	controlplanes, err := gatewayutils.ListControlPlanesForGateway(ctx, r.Client, gateway)
 	if err != nil {
-		return true, ctrl.Result{}, err
+		return false, ctrl.Result{}, err
 	}
 	if len(controlplanes) > 0 {
 		deletions, err := r.ensureOwnedControlPlanesDeleted(ctx, gateway)
 		if err != nil {
-			return true, ctrl.Result{}, err
+			return false, ctrl.Result{}, err
 		}
 		if deletions {
 			log.Debug(logger, "deleted owned controlplanes", gateway)
+			// Return early from reconciliation, deletion will trigger a new reconcile.
 			return true, ctrl.Result{}, err
```

Comment:
I know this was not part of your changes, but since we are here, can you return `nil` instead of the error? err here is always nil and it would increase the code's readability.

Reply:
```suggestion
			return true, ctrl.Result{}, nil
```